quilt/tools/command.py

```diff
@@ -884,29 +888,55 @@ def install(package, hash=None, version=None, tag=None, force=False):
                     print("Fragment already installed, but has the wrong hash (%s); re-downloading." %
                         file_hash)
 
-            response = s3_session.get(url, stream=True)
-            if not response.ok:
-                message = "Download failed for %s:\nURL: %s\nStatus code: %s\nResponse: %r\n" % (
-                    download_hash, response.request.url, response.status_code, response.text
+            temp_path_gz = store.temporary_object_path(download_hash + '.gz')
+            with open(temp_path_gz, 'ab') as output_file:
+                starting_length = output_file.tell()
+                response = s3_session.get(
+                    url,
+                    headers={
+                        'Range': 'bytes=%d-' % starting_length
+                    },
+                    stream=True,
+                    timeout=(S3_CONNECT_TIMEOUT, S3_READ_TIMEOUT)
                 )
-                raise CommandException(message)
 
-            length_remaining = response.raw.length_remaining
+                # RANGE_NOT_SATISFIABLE means, we already have the whole file.
+                if response.status_code != requests.codes.RANGE_NOT_SATISFIABLE:
+                    if not response.ok:
+                        message = "Download failed for %s:\nURL: %s\nStatus code: %s\nResponse: %r\n" % (
+                            download_hash, response.request.url, response.status_code, response.text
+                        )
+                        raise CommandException(message)
 
-            temp_path = store.temporary_object_path(download_hash)
-            with open(temp_path, 'wb') as output_file:
-                with tqdm(total=length_remaining, unit='B', unit_scale=True) as progress:
-                    # `requests` will automatically un-gzip the content, as long as
-                    # the 'Content-Encoding: gzip' header is set.
-                    # To report progress, however, we need the length of the original compressed data;
-                    # we use the undocumented but technically public `response.raw.length_remaining`.
-                    for chunk in response.iter_content(chunk_size=CHUNK_SIZE):
-                        if chunk: # filter out keep-alive new chunks
+                    # Fragments have the 'Content-Encoding: gzip' header set to make requests ungzip them
+                    # automatically - but that turned out to be a bad idea because it makes resuming downloads
+                    # impossible.
+                    # HACK: For now, just delete the header. Eventually, update the data in S3.
```

Comment:
Do you know of a way to update the data in s3 safely?

Reply:
Yeah, looks like it's doable: https://stackoverflow.com/questions/269840/is-it-possible-to-change-headers-on-an-s3-object-without-downloading-the-entire